### PR TITLE
Demolition derby event

### PR DESCRIPTION
**Pull request by Darxeal**
This event requires the Friendly Fire mutator.
All bots spawn at once at the specified `starts` - on random positions on a circle around the center.
The goal is to demo as many cars as you can, and avoid getting demoed.
If `perma_death` is true, when your bot is demolished, it won't respawn (it actually does but it gets teleported outside the map).
The event ends when there is only one bot alive or `max_duration` has passed.

### CODE REVIEW

File: events/demolition_derby.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
+"""
+Bot makers:
+
+To support this event type, you must listen on matchcomms for an event shaped like this (see DerbySpecification):
+{
+  "event_type": "DemolitionDerby",
+  "perma_death": true,
+  "max_duration": 60.0,
+  "starts": [
```
Comment (tarehart) [inline events/demolition_derby.py:9]: If I'm making a bot, should I ignore this and just look at the game tick packet during the countdown to see where everybody has spawned? From the message alone, I wouldn't know which index in "starts" was me.

File: events/demolition_derby.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    competitor_cfg_files: List[str]
+    result_demolitions: Dict[str, int]
+
+
+@dataclass
+class ActiveBotInfo:
+    competitor: Competitor
+    active_bot: ActiveBot
+    packet_index: int
+    time_lord: TimeLord
```
Comment (tarehart) [inline events/demolition_derby.py:71]: If each bot has a TimeLord, are we redundantly rendering text on screen? Not a big deal.

File: events/demolition_derby.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+        self.on_screen_log.log(f"About to spawn bots for DemolitionDerby.")
+        completed_spawns = self.spawn_helper.spawn_bots([competitor.bundle for competitor in self.competitors])
+
+        self.on_screen_log.log("Waiting for bots to get ready")
+        # We expect a number of ready messages equal to the number of competitors. Doesn't matter in which order they arrive.
+        # Currently we have no way of knowing which bot sent which message, so we don't know who supports this event.
+        for i in range(len(self.competitors)):
+            self.hide_ball()
+            self.spawn_helper.listen_for_events_supported_by_bot(timeout=7)
```
Comment (tarehart) [inline events/demolition_derby.py:150]: If we have 5 AdversityBots who never send ready messages, we'll be waiting for 35 seconds. I might tweak this so we sleep for a few seconds before the loop, and then use a minimal timeout in the listen for events call while looping.

**Review by tarehart — APPROVED** [review]
This is awesome, great event! I made some comments but I'm gonna go ahead and merge and we can think about those later.

File: events/demolition_derby.py (hunk 1 above)
Comment (Darxeal) [reply events/demolition_derby.py:9]: The index in the array should correspond with the index in the packet, but yeah it's not guaranteed. Maybe we can remove the start physics from the message. In waypoint race you can precompute your path or something, but here it's not very useful

File: events/demolition_derby.py (hunk 2 above)
Comment (Darxeal) [reply events/demolition_derby.py:71]: Good point, yeah this is a hacky solution for now

**Pull request merged**